### PR TITLE
[WIP] Include generators in Poly args

### LINKED ISSUE #9918: Several operations (simplify, Derivative, pickle) fail for constant Polys
Pickle succeeds to store Polys:

```
>>> x = Symbol('x')
>>> P = Poly(x-1)
>>> pickle.loads(pickle.dumps(P))
Poly(x - 1, x, domain='ZZ')
```

But it fails for constant Polys:

```
>>> x = Symbol('x')
>>> Q = Poly(1,x)
>>> pickle.loads(pickle.dumps(Q))
Traceback (most recent call last):
File "<pyshell#11>", line 1, in <module>
  pickle.loads(pickle.dumps(Q))
File "C:\Python34\lib\site-packages\sympy\polys\polytools.py", line 88, in __new__
  return cls._from_expr(rep, opt)
File "C:\Python34\lib\site-packages\sympy\polys\polytools.py", line 198, in _from_expr
  rep, opt = _dict_from_expr(rep, opt)
File "C:\Python34\lib\site-packages\sympy\polys\polyutils.py", line 332, in _dict_from_expr
  rep, gens = _dict_from_expr_no_gens(expr, opt)
File "C:\Python34\lib\site-packages\sympy\polys\polyutils.py", line 277, in _dict_from_expr_no_gens
  (poly,), gens = _parallel_dict_from_expr_no_gens((expr,), opt)
File "C:\Python34\lib\site-packages\sympy\polys\polyutils.py", line 238, in _parallel_dict_from_expr_no_gens
  raise GeneratorsNeeded("specify generators to give %s a meaning" % arg)
sympy.polys.polyerrors.GeneratorsNeeded: specify generators to give 1 a meaning
```

My guess is that the generator is dropped when pickling, and unpickling tries to evaluate `Poly(1)`, an expression which fails with the exact same error message when passed to the shell.

### LINKED ISSUE #15798: Poly.copy() does not copy unused generators
It looks to me like any generators that are not explicitly referenced in the a Poly expression  are dropped when performing a copy. This means `Poly.copy()` returns an unequal Poly object, which feels horribly wrong.

e.g.:
```
>>> import sympy
>>> from sympy.abc import x, y, z

>>> obj = sympy.Poly('x+y', x, y, z)
>>> o2 = obj.copy()
>>> obj == o2
False
>>> repr(obj), repr(o2)
("Poly(x + y, x, y, z, domain='ZZ')", "Poly(x + y, x, y, domain='ZZ')")
```
The same thing happens after pickling, leading to issues like those seen in https://github.com/sympy/sympy/issues/9918

In fact, it fails even the most basic:
```
>>> obj == obj.func(*obj.args)
False
```
Which should never happen : https://github.com/sympy/sympy/blob/28d913d3cead6c5646307ffa6540b21d65059dfd/sympy/core/basic.py#L704-L706

_Originally posted by @hackalog in https://github.com/sympy/sympy/issue_comments#issuecomment-455352417_

### PR DESCRIPTION
**Pull request by hackalog**
#### References to other Issues or PRs

Fixes #15798 
Fixes #9918 

#### Brief description of what is fixed or changed
`Poly.copy()` fails to produce an equivalent object if the polynomial expression doesn't include all the generators.  e.g.:
```
>>> import sympy
>>> from sympy.abc import x, y, z

>>> obj = sympy.Poly('x+y', x, y, z)
>>> o2 = obj.copy()
>>> obj == o2
False
>>> repr(obj), repr(o2)
("Poly(x + y, x, y, z, domain='ZZ')", "Poly(x + y, x, y, domain='ZZ')")
```

#### Other comments
This shows up in many ways, including constant polynomials (e.g. #9918)

In fact, it fails even the most basic:
```
>>> obj == obj.func(*obj.args)
False
```
Which should never happen : https://github.com/sympy/sympy/blob/28d913d3cead6c5646307ffa6540b21d65059dfd/sympy/core/basic.py#L704-L706

#### Release Notes

* polys
  * fix `copy()` issue when polynomial expression omits one or more generators

### CODE REVIEW

**Comment by oscargus** [comment]
Can you please include some tests? For example, the formerly not working example above.

   def test_poly_copy:
        obj = sympy.Poly('x+y', x, y, z)
        o2 = obj.copy()
        assert obj == o2

and something similar for the pickle example.

**Comment by jksuom** [comment]
It does not always suffice to give the generators, for example, if there are algebraic coefficients.
```
>>> p = Poly(x + sqrt(2), extension=True); p
Poly(x + sqrt(2), x, domain='QQ<sqrt(2)>')
>>> p.func(*p.args)
Poly(x + sqrt(2), x, domain='EX')
```

**Comment by hackalog** [comment]
I'm working on the test failures now. I'm not sure what to "correct" behavior of `atoms()` is supposed to be. These will clearly fail https://github.com/sympy/sympy/blob/28d913d3cead6c5646307ffa6540b21d65059dfd/sympy/core/tests/test_expr.py#L331-L332

as atoms just uses the `args()` output. With the new behavior, atoms will also return the generators that do not form part of the expression (or those that have been optimized out); i.e.

```
>>> Poly(1, x).atoms
{x, 1}
```
Should atoms return the generators that are not part of the expression?. e.g. should we see all the generators in cases https://github.com/sympy/sympy/blob/28d913d3cead6c5646307ffa6540b21d65059dfd/sympy/core/tests/test_expr.py#L337-L338

This is all related to many deeper assumptions about primitives in sympy than I'm comfortable guessing about.

**Comment by hackalog** [comment]
@jksuom Yes, I was worried about that, too. This is also a question I don't know the answer too, because unless `args()` returns the kwargs, too,  https://github.com/sympy/sympy/blob/28d913d3cead6c5646307ffa6540b21d65059dfd/sympy/core/basic.py#L704-L706 will never hold for such an example. (should this be a separate issue?)

**Comment by lmcinnes** [comment]
@jksuom Disconcertingly the equality testing seems to not care about the domain, but does care about the gens.

```python
>>> p = Poly(1 + x, x, domain='CC'); p
Poly(1.0*x + 1.0, x, domain='CC')
>>> q = p.copy(); q
Poly(1.0*x + 1.0, x, domain='RR')
>>> p == q
True
>>> r = Poly(1 + y, y, domain='CC'); r
Poly(1.0*y + 1.0, y, domain='CC')
>>> p == r
False
```

This is certainly not the behaviour I was anticipating. This was was in sympy 1.1.1, so perhaps upgrading will make it better?

**Comment by asmeurer** [comment]
`p` and `q` have different hashes, though. I don't understand why, because `Poly.__hash__` just reuses `Basic.__hash__`, which should just use the args.

Anyway, `Poly.__eq__` explicitly unfies the domains when checking equality, so I guess that behavior is intentional.

**Comment by hackalog** [comment]
I thought about this again last night. I think the fix has to be a little more drastic. The `Poly.args` output really needs to be a triple: `(expr, (gens), domain)`. This will require changing the constructor, as the gens are not put in a tuple, but passed as args directly. This will almost surely break some things. It's the only way to reconstruct everything on copy, pickle, ang generally, all the places where this is biting, while retaining some of the backwards compatibility (e.g.: infer the gens and domain from the expr, which is the current behavior).

Anything which relies on the old `Poly.args` behavior can easily be changed to use use `Poly.as_expr()`, since that's precisely what the old `Poly.args` was returning

**Comment by jksuom** [comment]
It might be possible to add domain to the tuple of gens. `Poly.__new__` would then test the last item of gens (assuming that there is at least one), and, if it is a domain (`isinstance(gens[-1], Domain)`), remove it from gens and save. It could then be added to `opt` replacing the domain possibly returned by `build_options`.

**Comment by asmeurer** [comment]
You mean add it to `.args`? I don't think it should go in `.gens`.

**Comment by jksuom** [comment]
I was thinking of how to deal with `Poly.__new__(cls, rep, *gens, **args)` if `domain` was to be appended to `Poly.args()`. That could be an extension to api that should not break anything. The keyword argument `domain=...` would still be preferred.

**Comment by asmeurer** [comment]
I think it could be done, since the domain is always a separate type that can't be a generator.

**Comment by hackalog** [comment]
Ok. I'll have a go at that...

**Comment by oscarbenjamin** [comment]
#15820 included a couple of tests that should apparently be uncommented when this PR is resolved. Maybe include that as part of this PR?

**Comment by oscarbenjamin** [comment]
Looks like this PR has some test failures that will need to be fixed...

**Comment by czgdp1807** [comment]
Closing this as both the issues are fixed in master. Please let us know if there is a need to re-open this. Thanks for your contributions.

**Comment by oscarbenjamin** [comment]
The issues are fixed on master from #18513 but no tests are added for pickle/copy so I still think the tests in this PR are useful

File: sympy/polys/polytools.py [hunk 1]
```diff
@@ -352,6 +352,26 @@ def free_symbols_in_domain(self):
...
+
+        >>> Poly(x**2 + 1, x).args
+        (x**2 + 1, x)
+
+        >>> Poly(x**2 + 1, x, y).args
+        (x**2 + 1, x, y)
+
+        """
+        return (self.as_expr(),) + self.gens
+
```
Comment (czgdp1807) [inline sympy/polys/polytools.py:374]:
```suggestion
```

**Comment by czgdp1807** [comment]
@oscarbenjamin I have retained the tests. Please see if this can be merged if tests pass.

File: sympy/polys/tests/test_polytools.py [hunk 2]
```diff
@@ -3323,6 +3325,29 @@ def test_issue_15669():
         2*2**Rational(4, 5)*x*(-x**2 + sqrt(8*x**2 + (x**2 - 2)**2) + 2)**Rational(3, 5) + 10*x)
     assert factor(expr, deep=True) == x*(x**2 + 2)
 
+def test_poly_copy():
+    # issue 15798
+    obj = Poly('x+y', x, y, z)
```
Comment (oscarbenjamin) [inline sympy/polys/tests/test_polytools.py:3330]: This should not use a string

File: sympy/polys/tests/test_polytools.py [hunk 3]
```diff
@@ -3323,6 +3325,29 @@ def test_issue_15669():
...
+    o2 = obj.copy()
+    assert obj == o2
```
Comment (oscarbenjamin) [inline sympy/polys/tests/test_polytools.py:3332]: This should explicitly assert that the generators are equal.

File: sympy/polys/tests/test_polytools.py (hunk 2 above)
Comment (oscarbenjamin) [reply sympy/polys/tests/test_polytools.py:3330]: same below

File: sympy/polys/tests/test_polytools.py (hunk 3 above)
Comment (oscarbenjamin) [reply sympy/polys/tests/test_polytools.py:3332]: same below

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/sympy/sympy/pull/19550

**Pull request closed without merging**